### ISSUE TITLE
Vsphere Cloud Provider: failed to detach volume from shutdown node

### ISSUE THREAD

**Issue by ksandermann** [kubernetes/kubernetes#75342]
**What happened**:
I have a pod with a pv attached to it running on node1.
When I shutdown node1 to simulate node-failure, Kubernetes detects the unhealthy node in the configured timeframe and tries to re-schedule the pod on node2 following the --pod-eviction timeout.
When trying to start the pod on node2, the pv can not be attached as it is still attached to the shutdown node1:
```
  Warning  FailedAttachVolume      6m               attachdetach-controller  Multi-Attach error for volume "pvc-db44144b-457f-11e9-a7b0-005056af6878" Volume is already exclusively attached to one node and can't be attached to another
```
Also, the pod on the shutdown node does not get deleted.

**What you expected to happen**:
The disk should be detached from the shutdown node and attached to the new node where the pod is scheduled on

**How to reproduce it (as minimally and precisely as possible)**:
1. Schedule single pod using pvc on node1
1. Shutdown node1
1. Watch FailedAttachVolume event on the new pod on node2

**Anything else we need to know?**:
Also, kube-controller-manager does not log anything about this failure.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.12.5
- Cloud provider or hardware configuration: vsphere
- OS (e.g: `cat /etc/os-release`): Ubuntu 16.04
- Kernel (e.g. `uname -a`): Linux k8s-dev-master3 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: Kubespray 2.8.3 using kubeadm
- Others: VSphere 6.5

**Comment by ksandermann** [comment]
@kubernetes/sig-vmware-bugs

**Comment by LinAnt** [comment]
It is even worse if you drain a node for an upgrade and then delete the VM. The disks that remain attached will be gone as well as the vm. This is not a recent issue, it has been like this since 1.8.x or 1.9.x.

**Comment by yastij** [comment]
There's a KEP opened for this kubernetes/enhancements#719

/priority important-soon

**Comment by ksandermann** [comment]
@yastij I see, thanks for the reference!
Is there any estimated time for this to actually get through? 
I see that the PR has been stale for 10 days now

**Comment by yastij** [comment]
The design is still in discussion, this will be landing on 1.15

**Comment by sjberman** [comment]
@ksandermann Do you also see an issue where the NotReady node never gets cleaned up? I'm seeing the issue you mentioned, but the powered down node stays in the NotReady state and never goes away. I'm wondering if it has anything to do with the fact that the pod with the volume attached still "exists" on the downed node (though in an "Unknown" state), and these two issues are somehow tied to each other.

**Comment by ksandermann** [comment]
@sjberman 

I didn't test that case, so I can't say anything to that.
What would your expected behaviour be in that case ? That k8s removes the NotReady Node completely from the cluster after time x ?

**Comment by sjberman** [comment]
@ksandermann Yeah, I would expect a node that is stuck in NotReady for some amount of time to be eventually removed from the cluster. Other cloud providers in different clouds have this behavior.

**Comment by zhonglin6666** [comment]
isMultiAttachForbidden return true When setting pv with accessMode value 'ReadWriteOnce'，
node1 attach and not detach when node1 down，more than one node attach volume，it will produce errors:
  attachdetach-controller  Multi-Attach error for volume "pvc-d0fde86c-8661-11e9-b873-0800271c9f15" Volume is already used by pod

```
// check if this volume is allowed to be attached to multiple PODs/nodes, if yes, return false
for _, accessMode := range volumeSpec.PersistentVolume.Spec.AccessModes {
	if accessMode == v1.ReadWriteMany || accessMode == v1.ReadOnlyMany {
		return false
	}
}
```

```
if rc.isMultiAttachForbidden(volumeToAttach.VolumeSpec) {
	nodes := rc.actualStateOfWorld.GetNodesForVolume(volumeToAttach.VolumeName)
	if len(nodes) > 0 {
		if !volumeToAttach.MultiAttachErrorReported {
			rc.reportMultiAttachError(volumeToAttach, nodes)
			rc.desiredStateOfWorld.SetMultiAttachError(volumeToAttach.VolumeName, volumeToAttach.NodeName)
		}
		continue
	}
}
```

**Comment by dfsdevops** [comment]
I'm also seeing this on 1.18.6, does anyone know any workarounds for this? fwiw I do not see any kind of "NotReady" nodes. I simply scaled the worker nodes down using the TKG cli, most pods got rescheduled successfully, others did not? How can I manually detach a volume?

**Comment by elmiko** [comment]
if this is still an issue, it should be reopened on the vsphere provider issues, https://github.com/kubernetes/cloud-provider-vsphere/issues , as the vsphere provider code has been removed from k/k.

/close
